service_discovery/core/src/registry_impl.rs

```diff
@@ -0,0 +1,457 @@
+};
+use std::collections::HashMap;
+use std::sync::Arc;
+use std::vec::Vec;
+use tonic::{Request, Response, Status};
+use tracing::{info, warn};
+
+/// Base structure for the registry gRPC service
+#[derive(Clone, Debug)]
+pub struct RegistryImpl {
+    pub registry_map: Arc<RwLock<HashMap<ServiceIdentifiers, ServiceMetadata>>>,
+}
+
+#[tonic::async_trait]
+impl Registry for RegistryImpl {
+    /// Registers a service by adding it to the registry.
+    ///
+    /// This function registers a service based on a [`RegisterRequest`]. Returns a
+    /// [`RegisterResponse`].
+    ///
+    /// # Arguments
+    ///
+    /// * `request` - A [`RegisterRequest`] wrapped by a [`tonic::Request`].
+    async fn register(
+        &self,
+        request: Request<RegisterRequest>,
+    ) -> Result<Response<RegisterResponse>, Status> {
+        let request_inner = request.into_inner();
+        let service_to_register =
+            request_inner.service.ok_or_else(|| Status::invalid_argument("Service is required"))?;
+        let service_identifiers = ServiceIdentifiers {
+            namespace: service_to_register.namespace.clone(),
+            name: service_to_register.name.clone(),
+            version: service_to_register.version.clone(),
+        };
+        info!("Received a register request for: {:?}", service_identifiers);
+
+        let registration_status;
+        // This block controls the lifetime of the lock.
+        {
+            let mut lock: RwLockWriteGuard<HashMap<ServiceIdentifiers, ServiceMetadata>> =
+                self.registry_map.write();
+            match lock.get(&service_identifiers) {
+                Some(_) => {
+                    lock.insert(service_identifiers.clone(), service_to_register.clone());
+                    registration_status = RegistrationStatus::Updated;
+                    info!("Updated the service entry in Chariott service registry; overwrote previous entry: {:?}", service_to_register);
+                }
+                None => {
+                    lock.insert(service_identifiers.clone(), service_to_register.clone());
+                    registration_status = RegistrationStatus::NewlyRegistered;
+                    info!(
+                        "Registered new service in Chariott service registry: {:?}",
+                        service_to_register
+                    );
+                }
+            };
+        }
+        let register_response =
+            RegisterResponse { registration_status: registration_status as i32 };
+        Ok(Response::new(register_response))
+    }
+
+    /// Unregisters a service by removing it from the registry.
+    ///
+    /// This function registers a service based on a [`UnregisterRequest`]. Returns a
+    /// [`UnregisterResponse`].
+    ///
+    /// # Arguments
+    ///
+    /// * `request` - A [`UnregisterRequest`] wrapped by a [`tonic::Request`].
+    async fn unregister(
```

Comment:
This doesn't need to be addressed now, but just wanted to bring this to light. 

Let's say an external provider is registered with Chariott, and there's a consumer that discovers this provider.

I am wondering if this consumer would have enough info from the discovery response to be able to unregister this external provider, or replace this provider in Chariott's registry using the register request?

It might be worth doing some extra validation in the future for unregister() and register().

Reply:
Agreed. I will create GitHub issues for this, but we need to investigate how we want to handle things like uniqueness and authorization for providers. How do we determine if another provider registering the same version of a service is a unique instance? How do we determine who can register/ unregister a service?

At the start, I see a basic introduction of something that determines if a Chariott service registry client can register or unregister. Over time this can develop by calling into some sort of AuthZ/AuthN service to make the determination or requiring a certificate or unique id. I think it's a great point that we need to investigate more in the future.